packages/runtime-common/realm-auth-handler.ts

```diff
@@ -0,0 +1,125 @@
+    this.realmURL = realmURL;
+  }
+
+  // A separate method for realm auth client creation to support mocking in tests
+  private createRealmAuthClient(
+    realmURL: URL,
+    matrixClient: MatrixClient,
+    loader: Loader,
+  ) {
+    return new RealmAuthClient(realmURL, matrixClient, loader);
+  }
+
+  fetchWithAuth = async (
+    request: Request,
+    retryOnAuthFail = true,
+  ): Promise<Response | null> => {
+    // todo rename method and retryOnAuthFail
+    if (request.url.startsWith(PACKAGES_FAKE_ORIGIN)) {
+      return null;
+    }
+
+    if (
+      request.url.endsWith('_session') ||
+      request.method === 'HEAD' ||
+      request.headers.has('Authorization')
+    ) {
+      return null;
+    }
+
+    let targetRealm: {
+      isPublicReadable: boolean;
+      realmAuthClient?: RealmAuthClient;
+      url: string;
+    };
+
+    let visitedRealmURL = Array.from(this.visitedRealms.keys()).find((key) => {
+      return request.url.includes(key);
+    });
+
+    if (visitedRealmURL) {
+      targetRealm = this.visitedRealms.get(visitedRealmURL)!;
+    } else {
+      let targetRealmHeadResponse = await this.loader.fetch(request, {
+        method: 'HEAD',
+      });
+
+      let targetRealmURL =
+        targetRealmHeadResponse.headers.get('x-boxel-realm-url');
+
+      if (!targetRealmURL) {
+        return null; // It doesn't look like we are talking to a realm (the request is for something else), so skip adding auth header
+      }
+
+      let isPublicReadable = Boolean(
+        targetRealmHeadResponse.headers.get('x-boxel-realm-public-readable'),
+      );
+
+      targetRealm = {
+        isPublicReadable,
+        url: targetRealmURL,
+        realmAuthClient: this.createRealmAuthClient(
+          new URL(targetRealmURL),
+          this.matrixClient,
+          this.loader,
+        ),
+      };
+
+      this.visitedRealms.set(targetRealmURL, targetRealm);
+    }
+
+    if (!targetRealm || !targetRealm.realmAuthClient) {
+      throw new Error(
+        `bug: should not have been able to get here without a visitedRealm or without an auth client`,
+      );
+    }
+
+    let isRequestToItself = targetRealm.url === this.realmURL; // Could be a request to itself when indexing its own cards
+
+    if (
+      isRequestToItself ||
+      (targetRealm.isPublicReadable && request.method === 'GET')
+    ) {
+      return null; // No need to add auth for GET to public readable realms
+    } else {
+      if (!this.matrixClient.isLoggedIn()) {
+        await this.matrixClient.login();
+      }
+      let jwt = await targetRealm.realmAuthClient.getJWT(); // This will use a cached JWT from the realm auth client or create a new one if it's expired or about to expire
+      request.headers.set('Authorization', jwt);
+
+      let response = await this.loader.fetch(request);
+
+      // 401 can mean the following: Missing token, expired token, malformed token, permissions changed (jwt payload does not match server permissions)
```

Comment:
I wonder if we should update the error response such that it communicates the specific situation where the permissions have changed, and thus could frequency of retries? Not a blocker for this PR.

Reply:
These 401 error responses actually do come with a response body:

1. `Authentication error: Missing 'Authorization' header`
2. `Authentication error: User permissions have been updated. Please refresh the token`
3. `Authentication error: Token expired`
4. `Authentication error: Invalid token`

I think (2) and (3) warrant doing a retry, while (1) and (4) indicate that something went wrong systemically

Reply:
https://github.com/cardstack/boxel/pull/1201/commits/99951c541430aa9d3d1a74e03151a4b645c38f61